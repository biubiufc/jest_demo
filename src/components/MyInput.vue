<template>
  <div>
    <input
      type="text"
      v-model="inputValue"
      maxlength="maxlength"
      @keyup.enter="handleEnter"
      @input="handleInput"
      @focus="show=true"
      @blur="show=false"
    />
    <ul v-show="show">
      <li v-for="item in historyRecord" :key="item">{{item}}</li>
      <li v-show="!historyRecord.length">暂无历史记录</li>
    </ul>
  </div>
</template>
<script>
 import store from "../store";
 import { mapActions } from 'vuex'
export default {
  data(){
    return{
      inputValue:this.value,
      show:false
    }
  },
  watch:{
    value(val){
      this.inputValue = val;
    },
  },
  computed:{
    historyRecord(){
      return store.state.records
    }
  },
  props: {
    value: {
      type: [String, Number],
    },
    maxlength: Number
  },
  methods: {
     ...mapActions(['incrementRecord']),
    handleEnter(event) {
      this.inputValue = ""
      if (!event.target.value) return;
      store.dispatch('incrementRecord',event.target.value)
      this.$emit("handleEnter", event.target.value);
    },
    handleInput(event){
      this.$emit("input", event.target.value);
    }
  }
};
</script>
<style lang="scss">
li{
  list-style: none;
}
li:nth-child(2n){
  background-color: #eee;
}
</style>
