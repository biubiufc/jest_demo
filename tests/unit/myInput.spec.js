import { shallowMount , createLocalVue  } from "@vue/test-utils";
import MyInput from "@/components/MyInput.vue";
import Vuex from 'vuex'
const localVue = createLocalVue()
localVue.use(Vuex)
describe("HelloWorld.vue", () => {
  let actions
  let store
  beforeEach(() => {
    actions = {
      incrementRecord: jest.fn()
    }
    store = new Vuex.Store({
      actions,
      state: {
        records: []
      },
    })
  })
  it("MyInput includes input", () => {
    const wrapper = shallowMount(MyInput);
    expect(wrapper.find("input").exists()).toBe(true);
  });

  it("MyInput default value is undefined", () => {
    const wrapper = shallowMount(MyInput);
    const vm = wrapper.vm;
    expect(vm.$props.value).toBe(undefined);
  });

  it("Myinput has maxlenth", () => {
    const wrapper = shallowMount(MyInput, {
      propsData: {
        maxlength: 2
      }
    });
    const vm = wrapper.vm;
    expect(vm.$props.maxlength).toBe(2);
  });

  it("No value pressing enter is invalid", () => {
    const wrapper = shallowMount(MyInput, {
      propsData: {
        value: ""
      }
    });
    const input = wrapper.find("input");
    input.trigger("keyup.enter");
    expect(wrapper.emitted().handleEnter).toBeFalsy();
  });

  it('enter after clear inputValue',async()=>{
    const wrapper = shallowMount(MyInput);
    const vm = wrapper.vm;
    const input = wrapper.find("input")
    await input.setValue('some value')
    await input.trigger("keyup.enter");
    expect(vm.$data.inputValue).toBe("")
  })

  it('snapShot',()=>{
    const wrapper = shallowMount(MyInput, {
      propsData: {
        value: 222
      }
    });
    expect(wrapper).toMatchSnapshot()
  })

 
 
  it("pressing enter record value",async()=>{
    const wrapper = shallowMount(MyInput,{ store, localVue });
    const input = wrapper.find('input')
    await input.trigger("keyup.enter");
    expect(actions.incrementRecord).toBeCalled()
  })
});

