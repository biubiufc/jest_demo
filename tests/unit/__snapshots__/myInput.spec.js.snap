// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld.vue snapShot 1`] = `
<div><input type="text" maxlength="maxlength">
  <ul style="display: none;">
    <li>some value</li>
    <li style="display: none;">暂无历史记录</li>
  </ul>
</div>
`;
